File: deepspace/src/deepspace/PowerEfficientSpaceStation.java
package deepspace;

public class PowerEfficientSpaceStation extends SpaceStation{
    private static final float EFFICIENCYFACTOR = 1.1F;

    public PowerEfficientSpaceStation(SpaceStation station){
        super(station);
    }

    @Override
    public float fire(){
        return super.fire()*EFFICIENCYFACTOR;
    }

    @Override
    public float protection() {
        return super.protection()*EFFICIENCYFACTOR;
    }

    @Override
    public Transformation setLoot(Loot loot){
        Transformation t=super.setLoot(loot);
        if(t==Transformation.SPACECITY) t=Transformation.NOTRANSFORM;

        return t;

    }

    @Override
    public PowerEfficientSpaceStationToUI getUIversion(){
        return new PowerEfficientSpaceStationToUI(this);
    }
}

File: PRACTICA4/ruby/PowerEfficientSpaceStation.rb


module Deepspace
  class PowerEfficientSpaceStation < SpaceStation
    @@EFFICIENCYFACTOR = 1.1

    def initialize(station)
      copy(station)
    end

    def fire
      return super * @@EFFICIENCYFACTOR
    end

    def protection
      return super * @@EFFICIENCYFACTOR
    end

    def setLoot(loot)
      t = super(loot)
      if t==Deepspace::Transformation::SPACECITY
        t=Deepspace::Transformation::NOTRANSFORM
      end
      return t
    end

    def getUIversion
      return PowerEfficientSpaceStationToUI.new(self)

    end

    def to_s
      return getUIversion.to_s
    end
  end
end